syntax = "proto3";
package flasherd;

service Flasherd {
  rpc RunCommand(stream RunCommandRequest) returns (stream RunCommandResponse);
}

// Represents a segment of a command. `regular` segments are passed as-is, while `binary` segments should correspond to a binary
// name previously sent. They will be populated with a host path to the streamed binary.
message Arg {
  optional string regular = 1;
  optional string binary = 3;
}

// Caller must start by streaming the binary to flash in chunks. The binary name should be consistent, and will be referenced
// in the arguments. Then, send a request with the command and args fields populated. Each subsequent Request packet may only
// contain stdin segments, which will be streamed to the command process.
message RunCommandRequest {
  optional string command_windows = 1;
  optional string command_macos = 2;
  optional string command_linux = 3;
  repeated Arg args = 4;
  optional bytes stdin = 5;
  optional string binary_name = 6;
  optional bytes binary_chunk = 7;
}

// Server will send zero or more packets with only stdout or stderr populated. When the command process terminates, the
// final packet sent will populate only the exit_code.
message RunCommandResponse {
  optional uint32 exit_code = 1;
  optional bytes stdout = 2;
  optional bytes stderr = 3;
}
